### PR TITLE
[SPARK-14541] [SQL] SQL function: IFNULL, NULLIF, NVL and NVL2

### PR DESCRIPTION
**Pull request by bomeng**
## What changes were proposed in this pull request?

I am trying to implement functions `NULLIF` in this PR. The meaning of NULLIF can be found here:
[NULLIF( )](https://oracle-base.com/articles/misc/null-related-functions#nullif)
[NVL( )](https://oracle-base.com/articles/misc/null-related-functions#nvl) / [IFNULL( )](https://www.1keydata.com/sql/sql-ifnull.html) `IFNULL()` and `NVL()` are the same.
[NVL2( )](https://oracle-base.com/articles/misc/null-related-functions#nvl2)
## How was this patch tested?

Test cases were added.

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/nullExpressions.scala [hunk 1]
```diff
@@ -128,6 +128,58 @@ case class IsNaN(child: Expression) extends UnaryExpression
...
+      null
+    } else {
+      valueLeft
+    }
+  }
+
+  override def genCode(ctx: CodegenContext, ev: ExprCode): String = {
+    val leftGen = left.gen(ctx)
+    val rightGen = right.gen(ctx)
+    dataType match {
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/nullExpressions.scala]: We have `CodegenContext.genEqual` to generate codes for equality check. You can see how `EqualTo` expression uses it.

Comment (bomeng) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/nullExpressions.scala]: Thanks, @viirya ! That simplifies the logic!

File: sql/core/src/main/scala/org/apache/spark/sql/functions.scala [hunk 2]
```diff
@@ -1015,6 +1015,33 @@ object functions {
...
+  def nullif(col1: Column, col2: Column): Column = withExpr { NullIf(col1.expr, col2.expr)}
+
+  /**
+   * Returns col2 if col1 equals is null, otherwise returns col1.
+   *
+   * @group normal_funcs
+   * @since 2.0.0
+   */
+  def nvl(col1: Column, col2: Column): Column = withExpr { Nvl(col1.expr, col2.expr)}
+  def ifnull(col1: Column, col2: Column): Column = withExpr { Nvl(col1.expr, col2.expr)}
```
Comment (davies) [inline sql/core/src/main/scala/org/apache/spark/sql/functions.scala]: docs

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/nullExpressions.scala [hunk 3]
```diff
@@ -128,6 +128,143 @@ case class IsNaN(child: Expression) extends UnaryExpression
...
+}
+
+/**
+ * An Expression accepts two parameters and returns the second parameter if the value
+ * in the first parameter is null; if the first parameter is any value other than null,
+ * it is returned unchanged.
+ */
+@ExpressionDescription(
+  usage = "_FUNC_(a,b) - Returns b if a is null, or a otherwise.")
+case class Nvl(left: Expression, right: Expression) extends BinaryExpression {
```
Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/nullExpressions.scala]: isn't this just coalesce?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/nullExpressions.scala [hunk 4]
```diff
@@ -128,6 +128,143 @@ case class IsNaN(child: Expression) extends UnaryExpression
...
+    }
+  }
+}
+
+/**
+ * An Expression accepts three parameters and returns the second parameter if the first parameter
+ * value is not null; if the first parameter is null, it returns the third parameter.
+ */
+@ExpressionDescription(
+  usage = "_FUNC_(a,b) - Returns b if a is null, or a otherwise.")
```
Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/nullExpressions.scala]: the example is wrong

File: sql/core/src/main/scala/org/apache/spark/sql/functions.scala [hunk 5]
```diff
@@ -1015,6 +1015,33 @@ object functions {
   def nanvl(col1: Column, col2: Column): Column = withExpr { NaNvl(col1.expr, col2.expr) }
 
   /**
+   * Returns null if col1 equals to col2, otherwise returns col1.
+   *
+   * @group normal_funcs
+   * @since 2.0.0
+   */
+  def nullif(col1: Column, col2: Column): Column = withExpr { NullIf(col1.expr, col2.expr)}
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/functions.scala]: maybe we don't need to add these to functions.scala. they are too sql specific

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/NullFunctionsSuite.scala [hunk 6]
```diff
@@ -55,6 +55,28 @@ class NullFunctionsSuite extends SparkFunSuite with ExpressionEvalHelper {
     checkEvaluation(IsNaN(Literal(5.5f)), false)
   }
 
+  test("NullIf") {
+    checkEvaluation(NullIf(Literal("a"), Literal("a")), null)
```
Comment (rxin) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/NullFunctionsSuite.scala]: need to test different data types

**Comment by rxin** [comment]
Also can you remove the "JIRA related" section from the description? It is redundant with the title when makes into the commit message.

**Comment by bomeng** [comment]
I will address these issues tomorrow! Thank you all!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/nullExpressions.scala (hunk 3 above)
Comment (bomeng) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/nullExpressions.scala]: I will say, yes, kind of. Here is what I found: [difference](http://stackoverflow.com/questions/950084/oracle-differences-between-nvl-and-coalesce)

Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/nullExpressions.scala]: we should document the difference if there are any

**Comment by bomeng** [comment]
I have revisited the codes and made the codes more robust. Heavily tested against different data types by using introducing testAllTypes2Values() with 2 different values. PR description was updated. Javadoc was fixed. Please leave comments!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/nullExpressions.scala (hunk 3 above)
Comment (bomeng) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/nullExpressions.scala]: Did not notice this. I will do it shortly.

**Comment by bomeng** [comment]
@rxin Could you please take a look if you get a chance? Thanks.

**Comment by rxin** [comment]
I added https://github.com/apache/spark/pull/13084

I thought about it more -- since it is unclear how often people will actually use these functions, it's probably not worth following strictly the corner case behaviors in Oracle/MSSQL. I created a new small framework that allows replacing the expressions at runtime with existing expressions, so we don't need to re-implement codegen.

**Pull request closed without merging**